### ISSUE TITLE
Internet access required ?

### ISSUE THREAD

**Issue by rickgitdone** [gtxaspec/wz_mini_hacks#38]
I installed the SD CARD made all the rtsp edits and booted up ... seems this still requires access to wyze for authorization before anything else works?

Is this possible to run local only on home subnet w/o any Internet access to the CAM?

**Comment by gtxaspec** [comment]
as far as I know, yes internet access is required.  I haven't done any testing without internet access at all.

**Comment by shamlord** [comment]
For setting up without the Wyze App, it may be possible. This thread has some details on the QR code setup process:
https://github.com/HclX/WyzeHacks/issues/146

As for offline use, I've been working to set up my four v3 cams with wz_mini_hacks today. Unintentionally, they did have internet access until just about 30 minutes ago. I didn't realize that adding another interface to my router for this second network would automatically created PASS rules to WAN until this thread prompted me to double check.

I have my cameras on a dedicated wifi network that has no access to my main network or the internet. I plan to grab the most recent updates tomorrow (RTSP sub streams? Awesome!) but I'll report back if they get into reboot loops. I'm the camera's currently have firmware 4.36.9.131 loaded internally.

**Comment by rickgitdone** [comment]
my hope is setup wifi with config file and launch the rtsp server v4l2rtspserver and ssh server ( dropbear) and leave iCamera and all the monitor processes down to allow local only system...  wired ethernet is good also ...

**Comment by gtxaspec** [comment]
currently wz_mini **_needs_** iCamera.  We hook the functions coded inside of iCamera to pass video to the rtsp server, so all the magic is done inside iCamera, without it nothing works.  You could look to the T31 SDK to program your own video program, but that is outside the scope of this project.

**Comment by shamlord** [comment]
Just following up on my isolated camera network experience.. The cameras functioned overnight after being cut off of IP or DNS access. This afternoon I updated the cameras to use the more recent build of wz_mini_hacks available earlier today, and afterwards they cycled between streaming RTSP and not. Mostly not. 

After giving the cameras IP and DNS access again, they stayed online.

If there is any way to spoof iCamera into not rebooting the cameras repeatedly without internet access that would be neat.

**Comment by gtxaspec** [comment]
I will run a test on a spare camera, and see if I can examine the logs to see what it uses as a check for internet access to prevent reboots.

**Comment by gtxaspec** [comment]
does anyone have an approximate time (hours?) of when the camera will reboot?

**Comment by gtxaspec** [comment]
quickly checking the logs...
```
May 16 16:42:45 iCamera: [init.c:565]gateway recovery ok...
May 16 16:42:46 iCamera: [DN:829]err: (getaddrinfo) fail:-2(Name or service not known), (domain: www.google.com)
May 16 16:42:46 iCamera: [netservice.c:896][test_conn_cloud_by_url]err: get_ip_addr_by_domain fail:1
May 16 16:42:46 iCamera: [DN:829]err: (getaddrinfo) fail:-2(Name or service not known), (domain: api.wyzecam.com)
May 16 16:42:46 iCamera: [netservice.c:896][test_conn_cloud_by_url]err: get_ip_addr_by_domain fail:1
May 16 16:42:48 iCamera: [DN:829]err: (getaddrinfo) fail:-2(Name or service not known), (domain: www.google.com)
May 16 16:42:48 iCamera: [netservice.c:896][test_conn_cloud_by_url]err: get_ip_addr_by_domain fail:1
May 16 16:42:48 iCamera: [DN:829]err: (getaddrinfo) fail:-2(Name or service not known), (domain: api.wyzecam.com)
May 16 16:42:48 iCamera: [netservice.c:896][test_conn_cloud_by_url]err: get_ip_addr_by_domain fail:1
```

so if anyone wants to try these commands via ssh:

`echo -e "127.0.0.1 www.google.com\n127.0.0.1 api.wyzecam.com" > /tmp/hosts ; mount --bind /tmp/hosts /etc/hosts`
 
and see if the camera stays up

**Comment by shamlord** [comment]
It's much less than an hour. If the camera starts up and does not have internet access, it seems to reboot almost constantly. 

I just cut off access and rebooted an easy to access camera. Once I can get into SSH on the camera I'll give this a shot.

**Comment by shamlord** [comment]
I can't seem to get the camera to even respond to pings, let alone allow me to SSH in. 

I just added the commands to the end of run_mmc.sh to see if that works.

**Comment by gtxaspec** [comment]
no pings even on LAN? hmm... can you share the run_mmc.log in the log dir when you have a chance?

**Comment by shamlord** [comment]
Its odd, after one power cycle the camera started responding to pings almost immediately. After another, it didn't start responding to pings for several minutes, responded to about 20 pings, then stopped responding again. I was able to SSH in long enough during another window of responsiveness, but the connection dropped before I could do anything. 

I just pulled the SD card and grabbed the requested log:
[run_mmc.log](https://github.com/gtxaspec/wz_mini_hacks/files/8704211/run_mmc.log)

**Comment by shamlord** [comment]
I'm looking through the log and if I read this correctly: 

`+ grep 'inet addr'
+ ifconfig wlan0
          inet addr:10.42.0.190  Bcast:10.42.0.255  Mask:255.255.255.0`

The camera is being configured with an IP not in my network's range.  I'm using 192.168.2.0/24. This specific camera has a DHCP reservation for 192.168.2.17, which does show up later in the logs.

**Comment by gtxaspec** [comment]
can you check your router? maybe there is a stale dhcp static lease or something.  perhaps you can check the logs too to see if and why it's giving out that 10.x ip.

**Comment by shamlord** [comment]
Its similar to an IP range I used years ago  (10.42.69.0/24 .. heh) but I currently use 172.16.0.0/16 for my home network and 192.168.2.0/24 for my camera network.  I have confirmed the ranges on my router's DHCP server (pfSense), and both of my Wifi APs are Asus routers in AP mode running the Merlin firmware. 

A DHCP diagnostic utility on a windows laptop connected to my camera Wifi is not getting replies, interestingly, so I'm troubleshooting that now.

**Comment by shamlord** [comment]
According to this forum thread, the cameras also fail if they cannot access specific NTP servers:
https://forums.wyzecam.com/t/wyze-cam-v3-connection-time-out-during-setup-when-ntp-being-forwarded/150908

Its been a few months, but this thread is interesting with regard to the camera failing to stay online based on DNS and NTP connection failures:
https://forums.wyzecam.com/t/wyze-cam-v3-connection-time-out-during-setup-when-ntp-being-forwarded/150908/3

**Comment by shamlord** [comment]
> so if anyone wants to try these commands via ssh:
> 
> `echo -e "127.0.0.1 www.google.com\n127.0.0.1 api.wyzecam.com" > /tmp/hosts ; mount --bind /tmp/hosts /etc/hosts`
> 
> and see if the camera stays up

I added this to the end of run_mmc.sh on one camera, it caused it to frequently drop its connection. The camera would drop 2-5 pings once every minute or so. The log file [run_mmc.log](https://github.com/gtxaspec/wz_mini_hacks/files/8713285/run_mmc.log) just shows Blue Iris repeatedly reconnecting. 

I removed this line and rebooted the camera, it hasn't dropped since. My other cameras have not been restarting, but were not blocked from WAN.

**Comment by gtxaspec** [comment]
this was for a camera with LAN access, but no internet right?

**Comment by shamlord** [comment]
Yes. It has the same behavior with internet access as well.

**Comment by gtxaspec** [comment]
i'll do more testing, but heres what i did:

start camera up with network, then changes dns server to 0.0.0.0

camera pings LAN, RTSP works, but no internet access in or out of the camera.

run ```echo -e "127.0.0.1 www.google.com\n127.0.0.1 api.wyzecam.com" > /tmp/hosts ; mount --bind /tmp/hosts /etc/hosts```

check 4 hours later, camera still online, rtsp server still working.

**Comment by shamlord** [comment]
On my network, the camera seemed to drop its connection every 15-20 seconds, but I think now that it does not actually reboot. My SSH connection would freeze, but then seemed to start responding again after a few moments. The connection would not drop, and whatever I was doing was not lost (i.e. nano was still running with run_mmc.sh open). Its possible the camera may be disabling and reenabling its interface or something similar. 

The way I blocked the camera's access was in the router, disallow everything from the camera network except the camera's access to the router's IP on that network (192.168.2.1). DNS would still be working, but the camera would not be able to make a ICMP, TCP or UDP connection to anything outside of the network. 

I can do some more troubleshooting tomorrow.

**Comment by shamlord** [comment]
There is a huge thread on the Wyze Forums about using the RTSP firmware and isolating the cameras from the Internet, what they try to access, and the experiences people have had trying to run the cams without Internet access:
https://forums.wyzecam.com/t/beta-testing-for-wyze-cam-v3-rtsp-firmware-now-available/196922/622

I tried the following so far today. The router interface the cameras are attached to is set to default DENY all traffic except for what I have allowed for each test. After making a change to my firewall rules, the camera was restarted with the `reboot` command via SSH. When the cameras disconnect, they are unreachable for about 3 to 5 seconds.

Test 1:
- Allow only UDP 123 from the cameras to WAN  (No DNS or anything else)
**Result: Cameras drop connection about every 60 to 70 seconds**


Test 2:
- Allow UDP 123 from cameras->WAN
- Allow All ports/protocols including DNS to the router's IP (but nothing to WAN)
**Result: Cameras drop connection about every 4 minutes**


FWIW, here are the NTP servers the camera tries to reach. I've not had success yet redirecting them to my router's NTP server via the hosts file on the camera. The hosts file does redirect the name resolution for the NTP servers but timesync fails to sync the time. The hostnames are hardcoded into the `/system/bin/timesync` binary:
```
clock.fmt.he.net
ntp0.cornell.edu
time-a.nist.gov
time.windows.com
time-b.nist.gov
```

- I pkilled the timesync process and restarted it to capture the following output, ending with me allowing NTP access again and the process syncing the time. [timesync.log](https://github.com/gtxaspec/wz_mini_hacks/files/8721424/timesync.log)


The camera has also tried to make connections to the following additional addresses based on the router/DNS log files of a few other users in the thread linked above:
```
api.wyzecam.com
wyze-membership-service.wyzecam.com
wyze-general-api.wyzecam.com
www.google.com
kinesisvideo.us-west-2.amazonaws.com - possibly for video processing or machine vision
wyze-iot.s3-us-west-2.amazonaws.com - retrieving/storing something in an S3 bucket
c1ybkrkbxxxxxx.credentials.iot.us-west-2.amazonaws.com
m-756xxxxx.kinesisvideo.us-west-2.amazonaws.com
r-fd5xxxxx.kinesisvideo.us-west-2.amazonaws.com
```

The addresses with the "xxxxx" at the beginning of the domain were partially redacted as the commenter was worried they might uniquely identify them.

I'm not super familiar with the Wyze cameras, but I'm learning as I go. If you have any insight that might help point me in a troubleshooting direction that would be neat. To get these camera's truly offline I think they will at least need to be redirectable to a local NTP server, and then their connections to other web services would need to be redirected to a lightweight web server (nginx, Lighttpd, etc) supplying the camera firmware with whatever response makes it happy.

**Comment by shamlord** [comment]
Something I forgot to add. According to Wyze, the cameras need the following access to the Internet according to Wyze's documentation:

<html>
<body>


Port | What it does | What it's used for
-- | -- | --
TCP: 80 | Local timelapse download | Timelapse
TCP: 123 | Internet time check | Confirming the camera's time zone
TCP: 443 | Cloud data transfer | Uploading Events
TCP: 8443 | Cloud API | Keeping the connection with Cloud server
TCP: 8605 | Upgrade package download | Firmware Upgrades
TCP: 10001 | P2P streaming connection | Local live streaming over WiFi
TCP: 10002 | LAN firmware upgrade | Firmware Upgrades
TCP: 22345 | TCP control | Making sure the connection stays open when viewing the Live Stream
UDP: 80 | Heart beat & streaming data | For all UDP ports: Loading the Live Stream and keeping the connection between camera and server even when it's not actively being used
UDP: 443 | Heart beat & streaming data |  


</body>
</html>

**Comment by shamlord** [comment]
Interesting.. I see that according to that table they want TCP 123 not UDP. Oof. Let me try that again..

**Comment by mrlt8** [comment]
Have you tried disabling all motion/sound detection/recording/notifications? Each of those events probably tries to phone home.

From what I can tell the problematic one is probably AWS IOT core (MQTT on 8443) which is also what all the unofficial wyze SDKs use to communicate with the camera over the wyze web api.

**Comment by mrlt8** [comment]
Actually, I wonder if `touch /tmp/aws_iot/rooCAok` might be enough to keep it from looping?

**Comment by capncybo** [comment]
So Just to clarify... Does this problem STILL affect the cameras EVEN if using them in WebCam mode???

**Comment by gtxaspec** [comment]
webcam mode disables everything. it's just a plain dumb webcam

**Comment by capncybo** [comment]
Well then, That's Fantastic... It's readily available, fairly in-expensive, & unlike most webcams has a speaker built-in.
Nearly Perfect for say a Raspberry PI or such a computer.
Although I'm not "YET" sure if the camera's speaker would be accessible via WebCam mode...

**Comment by rickgitdone** [comment]
have you tested web cam mode without internet access? if we could get the mode set and run the rtsp server we'd have an offline cam ...

**Comment by gtxaspec** [comment]
webcam mode does not work with rtsp as it does not run iCamera.

**Comment by rickgitdone** [comment]
in webcam mode does ssh also not come up?

process shows rtsp server using /dev/video1 
454     1 root     S    83144 88.8   0  1.9 /media/mmc/wz_mini/bin/v4l2rtspserver -C 1 -a S16_LE /dev/video1,hw:Loopback,0 -U noadmin:noadmin -P 8554

** when I look at the processes it looks like rtspserver relies on /dev/video1 to stream

I believe iCamera sets up the GPIO and some hardware during its startup... the kernel should load the devtree ( hopefully )

if we get the hardware up and /dev/video1 available we possibly can get the rtsp server to run?

**Comment by gtxaspec** [comment]
webcam mode switches the USB mode in the camera to client mode, which means no network access at all to the camera unless you have a dhcp server running on your PC, most people don't.   so no ssh .

webcam mode does not provide access to the video stream, so loading the loopback module for /dev/video1 won't work either.  simply put, webcam mode works as a plain USB video class webcam, only.

these camera's use a unique implementation for video using ingenic's libibmp.  iCamera uses the library to create 3 "framesource" nodes in /dev, /dev/framechan0 1 and 2.  these are locked by mutex to iCamera.  

iCamera does all the setup.  Without it, we have no video.  in webcam mode, there is no iCamera, but "uCamera".  uCamera does all the setup for the video and USB magic.

**Comment by gtxaspec** [comment]
I don't run my camera's offline, so I am not too familiar with what everyone is trying to achieve here.  What i understand is:

- Desire to run these cameras on a local LAN, with no internet access
- Access the RTSP streams on the camera from another device on the LAN

But, there are some blockers to achieving the above:
- the camera from reboots automatically after x amount of time due to lack of internet access
- packet loss due to lack of internet access

is this correct?

**Comment by mrlt8** [comment]
Has anyone tried `touch /tmp/aws_iot/rooCAok`? I'm pretty sure it's the camera trying to pull the  AWS certificates since those get removed and pulled at every boot.

**Comment by gtxaspec** [comment]
I'll test it tonight

**Comment by shamlord** [comment]
> I don't run my camera's offline, so I am not too familiar with what everyone is trying to achieve here. What i understand is:
> 
> * Desire to run these cameras on a local LAN, with no internet access
> * Access the RTSP streams on the camera from another device on the LAN
> 
> But, there are some blockers to achieving the above:
> 
> * the camera from reboots automatically after x amount of time due to lack of internet access
> * packet loss ( from cpu load?) due to lack of internet access
> 
> is this correct?

For me, this is correct.. though I do not know if the packet loss is due to CPU load or something else. 

There is some speculation on forum threads I've read about the camera possibly resetting its interface or rebooting when it doesn't have internet access. Those users would have been likely using the official Wyze RTSP beta firmware so the root cause may not be the same. 

It begs the question, why go through all the effort of cobbling together a RTSP solution out of a Wyze cam rather than just buying another camera that readily supports RTSP or ONVIF without all the trouble. For me, nothing matches the image quality and flexibility of the Wyze cameras in their tiny size. Last but not least, the window mount for these cameras is the number one reason why I chose the Wyze v3. I have to move with my job and they are so much easier to set up and take down than a bunch of wired bullet or dome cameras.

The problem is I do not want these cameras anywhere near the internet unless it is through my NVR and VPN.

**Comment by virmaior** [comment]
What's weird is that I've had my 5 V3s running without internet for months until this project came along.  So this must be version dependent (my cameras are on 4.36.2.5). So what version started generating this requirement?

**Comment by shamlord** [comment]
I don't know from personal experience which version started causing this. I would like to try older versions to see if they do not have this issue, but Wyze took them down.

If anyone can share their older firmware that might be one way to handle this.

**Comment by virmaior** [comment]
I can confirm that they keep recording and don't die without internet at least through 4.36.2.5 ...
**but** mine were running wyzeHacks and something breaks on versions of wz_mini_hacks after the 2022-05-12 where it says "waiting for wlan0 ..." (This could be an incompatibility arising from wzyeHacks + wz_mini_hacks )

I think versions after 4.36.3.18 have the new root password, which made wyzeHacks unusable, but this has no impact on wz_mini_hacks

**Comment by gtxaspec** [comment]
@virmaior, can you confirm on 4.36.2.5 cam had WLAN, just no internet access?

**Comment by virmaior** [comment]
yes most definitely. Isolated non-internet connected wifi wlan.

using wyze_hacks, I was downloading the video every hour to process (https://github.com/virmaior/zakkun-the-wyze). primary issue was having to manually set the time if the cameras ever rebooted. Been running this way since February and wyze's forced firmware attempts. Two cameras still not switched over to direct usb are still running this way right now.

obviously no access to the cameras via wyze app when running in non-internet connected mode.
they do seem to aggressively re-request DHCP leases so I've actually got my router restarting wifi every day at noon (my application is for a nocturnal flying squirrel so that's the time he won't be awake!)

**Comment by rickgitdone** [comment]
so I also downloaded 4.36.2.5 firmware and reloaded it as a downgrade and can confirm where I have blocked internet access and the Camera stays up... there are consistent times when ping responses goes from milliseconds to seconds and then back to milliseconds periodically however the Cam doesn't fully reboot how it does with newer firmware and wz_mini is running throughout.

Watching the stream on VLC I do see skips during the high ping response times and this seems to correlate when/where iCamera or whatever underlying process is trying to reach home and getting no response. so this version doesn't fully reboot however I am not calling this stable... its usable but not for any security purposed yet...

**Comment by gtxaspec** [comment]
so it looks like what iCamera is doing, if connectivity tests fail, it cycles the WiFi chip off and back on, probably explaining the packet loss you guys are seeing. 

if we replace wpa_supplicant with a fake, for example: ```touch /opt/wz_mini/bin/wpa_supplicant_fake; chmod +x mount --bind /opt/wz_mini/bin/wpa_supplicant_fake  /bin/wpa_supplicant;mount --bind /opt/wz_mini/bin/wpa_supplicant_fake  /bin/wpa_supplicant``` the wifi no longer gets cycles off and on.

**Comment by virmaior** [comment]
raising and lowering the interface explains why those cameras make so many DHCP requests. Looking at my records, I think they do this approximately every 200 seconds on one of my cameras every 578 seconds (oddly specific?)

**Comment by capncybo** [comment]
@virmaior Good Lord how in the world could WYZE engineering think so many DHCP requests were acceptable?
Not very WISE for WYZE Guys ;-)
@gtxaspec & thanks AGAIN for sorting all this out... you REALLY ARE making these cameras Better & Better.
+ I'm learning a ton via this project...
from Linux OS functionality & networking into Video encoding, decoding, streaming etc etc...  Really Neat Stuff.

**Comment by rickgitdone** [comment]
so I also downloaded 4.36.2.5 firmware and reloaded it as a downgrade and can confirm where I have blocked internet access and the Camera stays up... there are consistent times when ping responses goes from milliseconds to seconds and then back to milliseconds periodically however the Cam doesn't fully reboot how it does with newer firmware and wz_mini is running throughout.

Watching the stream on VLC I do see skips during the high ping response times and this seems to correlate when/where iCamera or whatever underlying process is trying to reach home and getting no response. so this version doesn't fully reboot however I am not calling this stable... its usable but not for any security purposed yet...

**Comment by rickgitdone** [comment]
I followed the idea to remount wpa_cli to a fake but 1st copied the real into /opt/wz_mini/bin/ so I could use it following...

touch /opt/wz_mini/bin/wpa_cli1
chmod +x /opt/wz_mini/bin/wpa_cli1
cp /bin/wpa_cli /opt/wz_mini/bin/
mount --bind /opt/wz_mini/bin/wpa_cli1 /bin/wpa_cli

blocked internet and seems pretty stable ... this is the best result so far ... except when dhcp renew comes... that's my only concern with this method ...  so any idea on the dhcp renew?

**Comment by virmaior** [comment]
@rickgitdone  I was under the impression that the hyper common dhcp renewals are because they are 
```
ip wlan0 down
ip wlan0 up 
```
. And since it's lower/raising the interface, it renews the dhcp during the re-up. But I could be wrong.

But maybe either
 you mean DHCP renewing on your wifi?  -> I just run a DHCP server on the raspberry pi that serves that subnet
OR
you mean there's a separate DHCP renewal thrashing going on even when it's not raising/lowering the interface?

**Comment by rickgitdone** [comment]
> @rickgitdone I was under the impression that the hyper common dhcp renewals are because they are
> 
> ```
> ip wlan0 down
> ip wlan0 up 
> ```
> 
> . And since it's lower/raising the interface, it renews the dhcp during the re-up. But I could be wrong.
> 
> But maybe either you mean DHCP renewing on your wifi? -> I just run a DHCP server on the raspberry pi that serves that subnet OR you mean there's a separate DHCP renewal thrashing going on even when it's not raising/lowering the interface?

I am wondering since we mounted a fake link to wpa_cli none of the standard methods to bounce the wifi link will work including dhcp requests  ... or am I wrong here?  I am monitoring dhcp requests following this and no request are coming from the WyzeCam ...

[virmaior](https://github.com/virmaior) noted the high dhcp requests above .. since I haven't been monitoring these I posted minimal/none following this adjustment

**Comment by virmaior** [comment]
@rickgitdone I have 5 V3s - all run 4.36.2.5 
three have been switched over to wz_mini and are on usb direct - no rampant DHCP requests
two are on Wyze_Hacks , using default wpa_supplicant, and have rampant DHCP requests

I'm guessing that switching the wlan out puts an end to the rampant requests (but cannot confirm expect based on adding up your evidence and mine).

**Comment by gtxaspec** [comment]
the camera does expect some output from wpa_cli... don't use a blank, try using `opt/wz_mini/bin/wpa_cli.sh`

as far as dhcp, try killing the dhcp client.  `pkill -f dhcp`

**Comment by gtxaspec** [comment]
Yes, the app is required for setup.  The camera can not  be used without the app for setup.

**Comment by gtxaspec** [comment]
any traction on this?  did anyone else try the wpa_cli.sh method i posted earlier?

**Comment by ss1lverman** [comment]
@gtxaspec - confirmed!   After doing both the wpa_cli and kill dhcp client fixes (with latest firmware ver for my v3), and then cutting internet access, there were no noticeable skips or reboots.   Used vlc to view rtsp stream.   Very dope!

**Comment by ss1lverman** [comment]
re: setting up without the app

Since we know settings are stored in the cam's filesystem (/etc/config/.user_config, /configs/.wifissid, /configs/.wifipasswd), what is stopping us from saving a minimal config that includes our custom ssid, pw, and some default cam settings like fps, res, etc. ?      Is some sort of key or checksum that prevents this?

**Comment by rickgitdone** [comment]
> any traction on this? did anyone else try the wpa_cli.sh method i posted earlier?
[root@WyzeV3:~]# uptime
 18:25:45 up 3 days, 23:18,  0 users,  load average: 1.98, 1.92, 1.95
 
I noticed today your mount binding was backwards...

I am using   mount --bind /opt/wz_mini/bin/wpa_cli.sh /opt/wz_mini/bin/wpa_cli

this maps the binary to the shell script and seems to be clearing the blips I saw prior...

**Comment by gtxaspec** [comment]
I see, so even with the wpa_cli mods, the process is still:

1. enable internet on lan 
2. power camera on
3. disable internet on lan
4. camera stays online on lan only

Is this correct?  or can #1 and 3 be removed?

**Comment by rickgitdone** [comment]
> I see, so even with the wpa_cli mods, the process is still:
> 
>     1. enable internet on lan
> 
>     2. power camera on
> 
>     3. disable internet on lan
> 
>     4. camera stays online on lan only
> 
> 
> Is this correct? or can 1 and 3 be removed?

3.  disable internet after knowing cam has been able to contact Wyze successfully ... I do this thru the phone app.. 

I tried cutting Internet access shortly after the CAM was up and that failed. I then waited 10 mins and so far all good. 

Another issue when you get a power blip it won't return until full power cycle and internet enabled access allows it to call home and register once...  we are getting closer...

**Comment by patkidon** [comment]
I still can't get mine to work. Can i verify the process? Sorry, I'm relatively new to this.

Running on Wyze Cam v3 on 4.36.9.139

1. Enable Internet on VLAN 
2. Plug in camera
3. Verify camera in Wyze app
4. SSH into camera
5. Input command "mount --bind /opt/wz_mini/bin/wpa_cli.sh /bin/wpa_cli"
6. Input command "pkill -f dhcp"
7. Disable Internet on VLAN

When I do this, I still get the random resets within 40-60 seconds. On VLC this is shown as a small freeze for about 2-5 seconds, but it maintains connection. In Blue Iris this completely resets the connection.

Also, I've noticed that when I disable internet on the VLAN, the Wyze app is still able to see the camera (even when I go on LTE on my phone), until I reboot the camera. Am I just not properly blocking access?

**Comment by rickgitdone** [comment]
@patkidon -  I rolled my Cam back to 4.36.2.5, this version and maybe others work better than newer ( mostly ) ...I rolled back after another said he had several up on this version w/o internet for a while.  Download links are posted.

**Comment by jsolarin** [comment]
I added the following to a custom script, /opt/wz_mini/startup.sh, (192.168.1.9 being my ntp server)

```
#!/opt/wz_mini/bin/bash

echo -e "127.0.0.1 localhost" > /tmp/hosts
echo -e "127.0.0.1 wyze-membership-service.wyzecam.com" >> /tmp/hosts
echo -e "127.0.0.1 test-upgrade-service.wyzecam.com" >> /tmp/hosts
echo -e "127.0.0.1 beta-upgrade-service.wyzecam.com" >> /tmp/hosts
echo -e "127.0.0.1 wyze-upgrade-service.wyzecam.com" >> /tmp/hosts
echo -e "127.0.0.1 test-device-api.wyzecam.com" >> /tmp/hosts
echo -e "127.0.0.1 beta-device-api.wyzecam.com" >> /tmp/hosts
echo -e "127.0.0.1 device-api.wyzecam.com" >> /tmp/hosts
echo -e "127.0.0.1 dcl-api.wyzecam.com" >> /tmp/hosts
echo -e "192.168.1.9 time-a-g.nist.gov" >> /tmp/hosts
echo -e "192.168.1.9 time-b-g.nist.gov" >> /tmp/hosts
echo -e "192.168.1.9 clock.fmt.he.net" >> /tmp/hosts
echo -e "192.168.1.9 dns3.cit.cornell.edu" >> /tmp/hosts

mount --bind /tmp/hosts /etc/hosts
```

This seems to help stop the camera from looping the interface restart when it doesn't have internet, but I think the iCamera binary still does a heartbeat every 3 minutes to the wyze servers and kills the network for a second.

I tried "mount --bind /opt/wz_mini/bin/wpa_cli.sh /bin/wpa_cli" but that doesn't do anything since the files are the same when using a wired ethernet adapter.

Unfortunately, I think the heartbeat is ssl so you'd have to MITM to try and spoof the heartbeat with a python http server.

**Comment by gtxaspec** [comment]
There's a way to hook the function in iCamera to stop it from cycling the network on firmware up to 4.36.9.139.  Working on it.  

Be advised I've discovered that blocking some of  those domains may lead to iCamera crashing.

**Comment by jsolarin** [comment]
That would be great, its terrible that they restart the network stack from the app.

Interestingly, with those domains blocked, the Wyze app still works when I open the camera back up to internet... I'm on 4.36.2.5 currently. No iCamera crash yet.

**Comment by jsolarin** [comment]
@gtxaspec any news on how to hook that function? I have found that if I unblock the internet from the camera, and let it do it's heartbeat, I can reblock the internet for a while before the camera "senses" it's blocked. After a random time (hours to days) however, the camera will again go into that network restart loop.

**Comment by gtxaspec** [comment]
@jsolarin 
try:

`touch /tmp/restart_wlan0.sh; mount --bind /tmp/restart_wlan0.sh /system/bin/restart_wlan0.sh`

does this prevent the camera from cycling the network?

**Comment by jsolarin** [comment]
Interesting...

So I had it echo out to a file to see if it was calling that script... And it is...

```
touch /tmp/restart_wlan0.sh
chmod u+x /tmp/restart_wlan0.sh
echo '#!/opt/wz_mini/bin/bash' > /tmp/restart_wlan0.sh                                          
echo 'echo "Attempted to restart network at `date`" >> /tmp/restartwlan.log' >> /tmp/restart_wlan0.sh
mount --bind /tmp/restart_wlan0.sh /system/bin/restart_wlan0.sh >> /opt/wz_mini/startupscript.log
```

But it still drops packets every few minutes when the log gets written to... So it's still restarting the network stack somehow?

**Comment by gtxaspec** [comment]
yes, its re-running the dhcp client.  working on hooking it

**Comment by jsolarin** [comment]
At least now we have timing,

Attempted to restart network at Sat Jul 23 04:22:37 UTC 2022
Attempted to restart network at Sat Jul 23 04:26:00 UTC 2022
Attempted to restart network at Sat Jul 23 04:29:23 UTC 2022
Attempted to restart network at Sat Jul 23 04:32:46 UTC 2022
Attempted to restart network at Sat Jul 23 04:36:09 UTC 2022
Attempted to restart network at Sat Jul 23 04:39:33 UTC 2022

**Comment by jsolarin** [comment]
So, I attempted to hook the /sbin/udhcpc binary after it aquired an IP address... but that just put the camera into a boot loop haha.

**Comment by jsolarin** [comment]
@gtxaspec any update on hooking the dhcp client restart?

**Comment by jsolarin** [comment]
I was able to find that the udhcpc has a default.script file that is called.

Tue Aug  9 05:34:21 UTC 2022 /usr/share/udhcpc/default.script deconfig
Tue Aug  9 05:34:27 UTC 2022 /usr/share/udhcpc/default.script bound
Tue Aug  9 05:37:38 UTC 2022 /usr/share/udhcpc/default.script deconfig
Tue Aug  9 05:37:41 UTC 2022 /usr/share/udhcpc/default.script bound
Tue Aug  9 05:41:01 UTC 2022 /usr/share/udhcpc/default.script deconfig
Tue Aug  9 05:41:04 UTC 2022 /usr/share/udhcpc/default.script bound

The problem is, the first function, deconfig, starts with ifconfig up... so I think something is downing the interface before this is called.

I tried hooking the file with a null file, but then the interface just went down and never came back up.

**Comment by rickgitdone** [comment]
> quickly checking the logs...
> 
> ```
> May 16 16:42:45 iCamera: [init.c:565]gateway recovery ok...
> May 16 16:42:46 iCamera: [DN:829]err: (getaddrinfo) fail:-2(Name or service not known), (domain: www.google.com)
> May 16 16:42:46 iCamera: [netservice.c:896][test_conn_cloud_by_url]err: get_ip_addr_by_domain fail:1
> May 16 16:42:46 iCamera: [DN:829]err: (getaddrinfo) fail:-2(Name or service not known), (domain: api.wyzecam.com)
> May 16 16:42:46 iCamera: [netservice.c:896][test_conn_cloud_by_url]err: get_ip_addr_by_domain fail:1
> May 16 16:42:48 iCamera: [DN:829]err: (getaddrinfo) fail:-2(Name or service not known), (domain: www.google.com)
> May 16 16:42:48 iCamera: [netservice.c:896][test_conn_cloud_by_url]err: get_ip_addr_by_domain fail:1
> May 16 16:42:48 iCamera: [DN:829]err: (getaddrinfo) fail:-2(Name or service not known), (domain: api.wyzecam.com)
> May 16 16:42:48 iCamera: [netservice.c:896][test_conn_cloud_by_url]err: get_ip_addr_by_domain fail:1
> ```
> 
> so if anyone wants to try these commands via ssh:
> 
> `echo -e "127.0.0.1 www.google.com\n127.0.0.1 api.wyzecam.com" > /tmp/hosts ; mount --bind /tmp/hosts /etc/hosts`
> 
> and see if the camera stays up

I used this above for timesync since I have been running OK on latest pull of wz_min_hacks  and  Wyze 4.36.2.5 with Internet blocked..

edit hosts file and bind it to /etc/hosts so that timesync pulls correct time from a local source on your LAN segment...

/opt/wz_mini/etc/hosts  to below
127.0.0.1       localhost
<local NTP>     clock.fmt.he.net
<local NTP>     ntp0.cornell.edu
<local NTP>     time-a.nist.gov
<local NTP>     time.windows.com
<local NTP>     time-b.nist.gov

mount --bind /opt/wz_mini/etc/hosts /etc/hosts
kill timesync

time now correct in Wyze rtsp stream

**Comment by kohrar** [comment]
I'm using wz_mini_hacks on the old RTSP beta (4.61.0.3) and the wifi connection drops for ~4 seconds every ~20 seconds after blocking all internet traffic to it. It looks like something's rerunning wpa_supplicant and udhcpc during the dropout periods.  

Tweaking the hosts file with the domains @rickgitdone mentioned as well as the wyze API domains seemed to have substantially reduced those dropouts for me to ~4 seconds every ~90 seconds.

How are you guys getting the iCamera log messages?

**Comment by rca** [comment]
I'm very interested in getting the network dropouts sorted out as well.  

Not sure how helpful this is, but running `strings` on the `iCamera` binary shows `ifconfig wlan0 down` buried in there somewhere: 

```
[root@cam:etc]# strings /system/bin/iCamera | grep ifconfig
ifconfig wlan0 0.0.0.0
ifconfig wlan0 10.42.0.1 netmask 255.255.255.0
ifconfig wlan0 down
```

I can see them dropping wlan0 to configure hostapd so that you could potentially initialize the camera by joining a self-hosted network ... or as a last ditch effort to get an internet connection.

If looking at the latter scenario; is there some way to replace `ifconfig` with a defanged version that does nothing if iCamera is running it?  Something like:

```bash
#!/bin/sh

parent_pid=`grep '^PPid:' "/proc/$$/status" | cut -f 2`

grep -q iCamera "/proc/$parent_pid/cmdline"

# if iCamera is running this; don't do anything
if [ "$?" -eq "0" ]; then
    exit 0;
fi;

# otherwise, let it rip
exec real_ifconfig "$@"
```

Thanks for this project; it's pretty awesome.  ❤️

**Comment by endertable** [comment]
Possible, but you may have to do some more doubles checks.  E.g., iCamera calls a script or other executable to restart the network, for some reason, sometimes it picks up iCamera vs the other script/executable as the PPID.  ifconfig is also needed to bring the network up, so in effect you could prevent yourself from rejoining the network.

Kinda risky messing with ifconfig, but as long as you use flagfiles for testing so that you allow altered version to run only one session and not after reboot, for just in case reasons, it *might* be worth experimenting.

The other thing is, after isolated tests, is this accomplishing what you are going after.  If not, there would be no reason to put such a risky bind into a production release where it could affect many others negatively.

It would be useful for you to run some of your tests and report back your findings...

**Comment by rca** [comment]
Agreed about the risk; I'm considering a different approach: look into defanging `iCamera`, which is the actual problem.

> but as long as you use flagfiles for testing so that you allow altered version to run only one session and not after reboot

Sounds good.  Can you point me to documentation that outlines how to do this?  I searched the repo and your comment above is the only reference to "flagfiles".

Thanks!

---

FWIW, I put the cameras on an IoT network that doesn't allow access to the public Internet, which is why I'm interested in getting this sorted out.  At the moment I'm seeing ~1 camera dropout per hour based on Blue Iris's "No Signal" camera stat, which feels better than what I've observed before, but it would be nice to get signal drops down to 0 if possible.  I updated my firewall to allow NTP access, and that seems to have made the dropouts less frequent.

**Comment by endertable** [comment]
The following snippet would have to go right before iCamera is kicked off.   First read ASSUMPTIONS then EXPLANATIONS

======================================================
if [ -f /configs/TESTFLAG ]
then
    ##  Flag file exists so good to go on defanged ifconfig for this session
    mkdir /run/bin
    cp /opt/wz_mini/test/ifconfig_dfanged   /run/bin/ifconfig
    export PATH=/run/bin:$PATH
    ##  Kill the flagfile so if we reboot the cam, everything back to normal
    rm  /configs/TESTFLAG
else
    ##  No flagfile, cam must have restarted so do not use defanged ifconfig
    :
fi
======================================================

ASSUMPTIONS:
  You have your defanged script in /opt/wz_mini/test/ named 'ifconfig_dfanged'
  You create the flagfile when ready to test by executing:   'touch /configs/TESTFLAG'

EXPLANATIONS:
  In order for the script to run your test code, the flagfile '/configs/TESTFLAG' must exist, hence why you must create it.
  Once created, and above snippet inserted, reboot cam, and because flag file exists defanged ifconfig will be in PATH.
      Flagfile will also be immediately deleted in case something goes wrong, you can just reboot cam and it will start as normal
      until you create the flagfile again.
  Ensure your script has alot of verbose writing to a file on the SD card so you can analyze what you may need to revise in your
       script.
  If my memory serves me, specifically to ifconfig, something calls ifconfig using '/sbin/ifconfig' in whcih case, altering the path
       like above won't work, you will have to bind the whole /sbin directory to /tmp or /run    :(

OTHER POINTS:
  You can alter above snippet to whatever you want, and to make things easier, you *might* be able to put your defanged script
       into wz_mini's bin directory and just name it 'ifconfig', I just don't remember where in $PATH it is, but it needs to precede
      /sbin
  I use flagfiles for debugging, and other stuff, so if sh*t hits the fan you're not stuck in a bad loop.  Of course, worst case
     scenario, you can pull the SD card and alter directly, but who wants to do that?

Good luck, any other questions, just ask..

**Comment by kohrar** [comment]
I poked around in the `iCamera` binary and I seemed to have made it not bounce the network connection after being unable to reach the cloud servers or google.com. There're a few function calls that are made during the network initialization phase. By bypassing the calls to check whether the domain name is working and making the retry loop indefinitely to prevent it from resetting the network on failure, `iCamera` no longer bounces the network interface.

So far, I have the camera streaming the RTSP feed with absolutely no dropouts anymore (0% ping loss after a couple hours). I collected some [notes on this here](https://leo.leung.xyz/wiki/Wyze_Cam#Running_without_internet).

Here's the patch I came up with for the RTSP beta firmware 4.0.61.1. Simply log in to the Wyze Cam and run it. It will place the patched binary on the SD card and tweak the iCamera shim script to call the patched binary. No other modifications should be necessary to get this working. If you need to revert, you will have to copy back the original `iCamera` script to `wz_mini/usr/bin/iCamera` on your SD card.

I'll look at iCamera on the latest firmware when I have time.

```
#!/bin/sh
# See: https://leo.leung.xyz/wiki/Wyze_Cam#Running_without_internet
set -e

cd /tmp

# Make a copy to patch
cp /system/bin/iCamera iCamera

MD5Sum=$(md5sum iCamera | awk '{print $1}')
if [[ "$MD5Sum" != "04b90d6d77be72a4dd8c18da4b31946a" ]] ; then
        echo "This patch is for iCamera in firmware 4.61.0.1"
        exit 1
fi


# Patch out the calls to test cloud url, which calls that DN function
#  This isn't strictly necessary, though it will cause iCamera to constantly retry the tests and spam the iCamera outputs with messages like
#    DN:854]err: (getaddrinfo) fail:-2(Name or service not known), (domain: www.google.com
#  so we'll comment these calls here to make it hush up
function patch_out_calls_to_test_cloud_url() {
        for i in  0x603b0  0x602d4 ; do
                echo "Original at $i"
                dd if=iCamera bs=1 count=4 skip=$(($i)) 2>/dev/null | xxd

                echo "Patched"
                printf '\x00\x00\x00\x00' | dd conv=notrunc of=iCamera bs=1 seek=$(($i)) 2> /dev/null
                dd if=iCamera bs=1 count=4 skip=$(($i)) 2>/dev/null | xxd
        done
}


# Patch out the threadpool_add_job calls for net-valid and upload-rebootlog, and dongle send
function patch_out_jobs_after_connect () {

        # Everything up until the last branch instruction
        for i in $(seq 0x0070d0 4 0x07114) ; do
                echo "Original at $i"
                dd if=iCamera bs=1 count=4 skip=$(($i)) 2>/dev/null | xxd

                echo "Patched"
                # noop
                printf '\x00\x00\x00\x00' | dd conv=notrunc of=iCamera bs=1 seek=$(($i)) 2> /dev/null
                dd if=iCamera bs=1 count=4 skip=$(($i)) 2>/dev/null | xxd
        done
}


# When our calls to DN check cloud url fails, we run into the code that calls funky_network_function(0). We don't want that. So noop all this out
# All the code that sets DAT_005e4d54=0, prints debug message, and calls funky_network_func(0) is nooped out here.
# This fixes the network from going back to the idle state and bouncing everything
function patch_out_network_reset_to_idle () {
        for i in $(seq 0x006041c 4 0x06045c) ; do
                echo "Original at $i"
                dd if=iCamera bs=1 count=4 skip=$(($i)) 2>/dev/null | xxd

                echo "Patched"
                # noop
                printf '\x00\x00\x00\x00' | dd conv=notrunc of=iCamera bs=1 seek=$(($i)) 2> /dev/null
                dd if=iCamera bs=1 count=4 skip=$(($i)) 2>/dev/null | xxd
        done
}

patch_out_calls_to_test_cloud_url
patch_out_jobs_after_connect
patch_out_network_reset_to_idle


# Place it on the SD card
cp iCamera  /opt/wz_mini/usr/bin/iCamera.patched

# Patch the iCamera script to use the patched executable
sed -i 's/\/system\/bin\/iCamera/\/opt\/wz_mini\/usr\/bin\/iCamera.patched/' /opt/wz_mini/usr/bin/iCamera

```

**Comment by capncybo** [comment]
@kohrar This is fantastic work & very much appreciated, (much of your knowledge is far beyond my level of understanding).
I'm currently running firmware 4.36.9.139 But if I understood you correctly...
#1) Your patch is already working however, it's specifically for: RTSP beta firmware 4.0.61.1
#2) Sometime soon, you will attempt to patch firmware 4.36.9.139...
I hope I understood that correctly, Anyways:
Thank you for sharing your knowledge & expertise, & I wish you the best of luck & success in all your endeavors.

**Comment by gtxaspec** [comment]
@kohrar awesome work.  Are the functions you are patching native to the iCamera binary, or an external library? if they are external, we could probably add your fixes into libcallback via ld preload.

**Comment by kohrar** [comment]
Thanks for the complements and I'm glad that this is of interest to you all.

@capncybo, Yes, the patch is only for 4.0.61.1 (the RTSP beta). I'll look at the latest firmware this weekend.
@gtxaspec, I'm modifying the iCamera binary directly. The network reset is triggered there rather than some outside library, I think. Since all the tools necessary to do the patching is already part of wz_mini_hacks, perhaps this could just be bundled as a script that gets triggered when a config option is present.

**Comment by rca** [comment]
I have been looking into ways to get iCamera to call defanged versions of system executables, and I got it to use custom versions of `ifconfig`, `killall`, and `udhcpc`, but your way of doing this is much more elegant. 

The thought of writing a script that does this in boot was something I was going to look into upon reading your notes (nice notes, btw!).

I wasn't clear on how you went from determining the function names (presumably with the debug logs when you run the program in the foreground?) to file seeks within the binary.

Can you explain your procedure a bit here?  Perhaps I can help with the automation once I understand the mechanics. 

Thanks!!

**Comment by kohrar** [comment]
@rca Yes, that was my initial approach as well too. Looking at the iCamera binary though, I think it may be calling other things aside from those utilities that may be bouncing the network connection. 

The function names are just names I made up as I looked through the disassembly in Ghidra. As a result, the comments in my patch script probably probably won't make much sense. If you want to take a look, here's the Ghidra export which has some of the functions named with some bookmarks to the patch locations: [iCamera.gz](https://github.com/gtxaspec/wz_mini_hacks/files/9590930/iCamera.gz) Just rename it with a `.gzf` extension and import it into a new project.

--- 

Here's the updated patch script which should also work on the latest 4.36.9.139 firmware. The code I'm targetting seems to be mostly unchanged so only the addresses/offsets need to be changed. I tested this on a new out-of-the-box Wyze Cam running this firmware with wz_mini_hacks and I've had 0 ping drops after almost 2.5 hours. I'm also streaming RTSP using the v4l2rtspserver provided by wz_mini_hacks without any glitching (which was the reason why I was using the RTSP beta in the first place). As a side note, I had to add in a patch to get rid of the code_test stuff from triggering. For some reason this is being triggered which causes iCamera to poll the contents of `/tmp` every 10 seconds. This wasn't an issue on the RTSP beta.

I'll be upgrading all my other Wyze Cams from the FiveLeavesLeft hack to this firmware over the next few days. I hope someone can give this a try and give any feedback.

**Usage:** 
1. Copy this script to `/tmp/patch.sh` on the Wyze Cam v3. 
2. Execute the script by calling `bash patch.sh` (you have to use bash due to some bash specific syntax). 
3. Verify that the patch executed properly. The MD5 hash of my patched iCamera binary is `816a42f5f35dd41ad9aae19163e98c49`. 
4. Reboot. 
5. When the camera comes back online, you should see `iCamera.patched` running when you run `ps`.  

If you need to revert, you will have to copy back the original iCamera script to `wz_mini/usr/bin/iCamera` on your SD card and delete the `iCamera.patched` binary.

Enjoy!

```
#!/opt/wz_mini/bin/bash
set -e

cd /tmp

# Make a copy to patch
cp /system/bin/iCamera iCamera

Version=""

MD5Sum=$(md5sum iCamera | awk '{print $1}')
if [[ "$MD5Sum" == "04b90d6d77be72a4dd8c18da4b31946a" ]] ; then
        Version="4.61.0.1"
fi

if [[ "$MD5Sum" == "b1c96d966226d76db86c96ecdfdd79e9" ]] ; then
        Version="4.36.9.139"
fi

if [ -z "$Version" ] ; then
        echo "Unsupported version. MD5sum: $MD5Sum";
        exit 1
fi

# Patch out the calls to test cloud url, which calls that DN function
#  This isn't strictly necessary, though it will cause iCamera to constantly retry the tests and spam the iCamera outputs with messages like
#    DN:854]err: (getaddrinfo) fail:-2(Name or service not known), (domain: www.google.com
#  so we'll comment these calls here to make it hush up
function patch_out_calls_to_test_cloud_url() {
        [[ "$Version" == "4.61.0.1" ]] && Address="0x603b0 0x602d4"
        [[ "$Version" == "4.36.9.139" ]] && Address="0x89938 0x89858"

        echo -e "\n\n====> Calling ${FUNCNAME[0]}\n"

        for i in  $Address ; do
                echo -e "\nOriginal at $i"
                dd if=iCamera bs=1 count=4 skip=$(($i)) 2>/dev/null | xxd

                echo "Patched"
                printf '\x00\x00\x00\x00' | dd conv=notrunc of=iCamera bs=1 seek=$(($i)) 2> /dev/null
                dd if=iCamera bs=1 count=4 skip=$(($i)) 2>/dev/null | xxd
        done
}


# Patch out the threadpool_add_job calls for net-valid and upload-rebootlog, and dongle send
function patch_out_jobs_after_connect () {
        [[ "$Version" == "4.61.0.1" ]]   && Address="$(seq 0x070d0 4 0x07114)"
        [[ "$Version" == "4.36.9.139" ]] && Address="$(seq 0x7b184 4 0x7b1cc)"

        echo -e "\n\n====> Calling ${FUNCNAME[0]}\n"

        # Everything up until the last branch instruction
        for i in $Address ; do
                echo -e "\nOriginal at $i"
                dd if=iCamera bs=1 count=4 skip=$(($i)) 2>/dev/null | xxd

                echo "Patched"
                # noop
                printf '\x00\x00\x00\x00' | dd conv=notrunc of=iCamera bs=1 seek=$(($i)) 2> /dev/null
                dd if=iCamera bs=1 count=4 skip=$(($i)) 2>/dev/null | xxd
        done
}


# When our calls to DN check cloud url fails, we run into the code that calls funky_network_function(0). We don't want that. So noop all this out
# All the code that sets DAT_005e4d54=0, prints debug message, and calls funky_network_func(0) is nooped out here.
# This fixes the network from going back to the idle state and bouncing everything
#  The call to print debug message could probably be left intact... we don't actually prevent 005e4d50=0  though..
function patch_out_network_reset_to_idle () {
        [[ "$Version" == "4.61.0.1" ]]   && Address="$(seq 0x6041c 4 0x6045c)"
        [[ "$Version" == "4.36.9.139" ]] && Address="$(seq 0x899a4 4 0x899e4)"

        echo -e "\n\n====> Calling ${FUNCNAME[0]}\n"

        for i in $Address ; do
                echo -e "\nOriginal at $i"
                dd if=iCamera bs=1 count=4 skip=$(($i)) 2>/dev/null | xxd

                echo "Patched"
                # noop
                printf '\x00\x00\x00\x00' | dd conv=notrunc of=iCamera bs=1 seek=$(($i)) 2> /dev/null
                dd if=iCamera bs=1 count=4 skip=$(($i)) 2>/dev/null | xxd
        done
}


# For some reason on 4.36.9.139, iCamera starts the code_test section which just constantly lists /tmp over and over every 10 seconds.
# I don't know why this is getting triggered, so I'm going to patch this out from being called.
function patch_out_code_test_enable() {
        [[ "$Version" == "4.61.0.1" ]]   && return;  # No need to do this as it doesn't seem to be a problem
        [[ "$Version" == "4.36.9.139" ]] && Address="0x7dfcc"

        echo -e "\n\n====> Calling ${FUNCNAME[0]}\n"

        for i in $Address ; do
                echo -e "\nOriginal at $i"
                dd if=iCamera bs=1 count=4 skip=$(($i)) 2>/dev/null | xxd

                echo "Patched"
                # noop
                printf '\x00\x00\x00\x00' | dd conv=notrunc of=iCamera bs=1 seek=$(($i)) 2> /dev/null
                dd if=iCamera bs=1 count=4 skip=$(($i)) 2>/dev/null | xxd
        done
}

patch_out_calls_to_test_cloud_url
patch_out_jobs_after_connect
patch_out_network_reset_to_idle
patch_out_code_test_enable

echo -en "\n\nPatching done. MD5 hash: "
md5sum iCamera


# Place it on the SD card
cp iCamera  /opt/wz_mini/usr/bin/iCamera.patched

# Patch the iCamera script to use the patched executable
sed -i 's/\/system\/bin\/iCamera/\/opt\/wz_mini\/usr\/bin\/iCamera.patched/' /opt/wz_mini/usr/bin/iCamera
```

**Comment by rca** [comment]
Awesome, thanks for the update @kohrar !  I've patched my camera and will let it run overnight.

I'm comparing this to another camera that is running `4.36.9.139` + wz_mini_hacks.  With my attempt to patch network system executable calls I got the signal drops from 690 times down to 386 over ~2 days.  About half isn't bad, but not great either.  This is probably due to the network restarts getting defanged, but I did notice that something somewhere is rebooting the camera after some other failures.

I'm super excited about this approach and will report back.

Thanks again!  🙌🏽

**Comment by johnsolarino** [comment]
I also installed version 4.36.9.139 on 2 cameras and patched the iCamera binary. So far they haven't have any drops or packet loss without internet. Other cameras have 40 - 100 no signal alarms in the same time period. This looks great so far!

**Comment by jsolarin** [comment]
I also installed version 4.36.9.139 on 2 cameras and patched the iCamera binary. So far they haven't have any drops or packet loss without internet. Other cameras have 40 - 100 no signal alarms in the same time period. This looks great so far!

**Comment by rca** [comment]
After running `iCamera.patched` overnight it's had 0 drops compared to 60 on an unpatched camera; nice work!  🙌🏽

**Comment by plloppii** [comment]
Trying to test this as well, 
I'm struggling to turn off internet access for the wyze and getting it connected to my computer through an unmanaged switch. 
I'm using the V3 with the UCTRONICS POE ethernet adapter.  Any ideas?

**Comment by capncybo** [comment]
@kohrar Sorry for being late to the "Patch" party ;-)
I did successfully patch firmware v4.36.9.139 on THREE of my WYZE v3 Cameras.
  Fantastic Work BTW
From my perspective, It seemed to work great for the short-time I was using it.
However I quickly realized...
 from a WYZE Application perspective... the patch (obviously) OFF-LINES the cameras.
 Not sure if the wife & kids will like loosing their WYZE based access
 And I was also using the docker-wyze-bridge via https://github.com/mrlt8/docker-wyze-bridge
 which also breaks without the WYZE connectivity.
Hence I reverted (rather quickly) before the rest of the family even noticed. LOL

I did see your note about reverting...
 "If you need to revert, you will have to copy back the original iCamera script to wz_mini/usr/bin/iCamera on your SD card and delete the iCamera.patched binary."

But I was somewhat uncertain as to what file, & weather I still had the original iCamera script file elsewhere....
So I quickly decided to (try/run):
upgrade-run.sh
FWIW this did revert all functionality, as expected.

Anyways I do thank you for your excellent work.
But I guess I'll have to decide... WYZE Connectivity or Total Independence.
 Hmmm

**Comment by rickgitdone** [comment]
where is the best location to pull this patched code with git?

I ask since I am seeing a forked repo with the code and the discussion is here....

**Comment by rca** [comment]
@capncybo you are correct; the patch makes the cameras offline only.  Good call that there are some folks that do not want to completely disable the Wyze application.  Personally, that's exactly what I'm looking for, but this patch should be compatible with both use-cases.

The original, untouched binary is at `/system/bin/iCamera`.  You would have to update the `iCamera` wrapper script to point back to the original location; essentially the reverse of the last line in [the script in the comment above](https://github.com/gtxaspec/wz_mini_hacks/issues/38#issuecomment-1250000186):

```
sed -i 's/\/opt\/wz_mini\/usr\/bin\/iCamera.patched/\/system\/bin\/iCamera/' /opt/wz_mini/usr/bin/iCamera
```

Hope this helps.

**Comment by capncybo** [comment]
@rca Thanks for the added information, & yes indeed it did help :+1:

**Comment by kohrar** [comment]
@rickgitdone You can use the pull request #298. My fork's master branch also has this change, but I hope that this can eventually get merged in.

To @capncybo's point, perhaps the option for this feature should be called something to the effect of `ENABLE_SELFHOSTED_MODE` which should make it clear it will break the functionality with Wyze's cloud service. It didn't occur to me it would break Wyze functionality as I don't have any intention of using it, but normal users that intend to use it with internet connectivity should not be using this patch anyway since their wireless connection wouldn't be dropping intermittently. This is why the feature flag is disabled by default in my PR.

**Comment by gtxaspec** [comment]
@kohrar Thank's for the pull requests!  I don't have access to a camera to test for the next few days, but once I do, I'll get on merging these in ASAP!

**Comment by xraive** [comment]
@kohrar Would you have time to make a patch for Wyze Cam v2? The latest firmware version is 4.9.8.1002.

**Comment by rickgitdone** [comment]
I changed the firmware over to v4.36.9.139 to match the iCamera binary being patched... pulled the #298 today edited the  ENABLE_SELFHOSTED_MODE to "true"

don't see my CAM coming on WIFI now ... without the SDCard in the CAM is working... might there be an issue in the forked repo ? I did wipe the SDCard in between so that its fresh each time as well...

I created a new clone of https://github.com/kohrar/wz_mini_hacks.git  so that the orig and this fork are separate that I could go back and forth for testing...

**Comment by kohrar** [comment]
@rickgitdone, Could you please send the wz mini boot logs? It should be on your SD card. Perhaps it would be more fitting to comment directly on my PR #298 to keep everything related to the change there. Thanks!

**Comment by kohrar** [comment]
> @kohrar Would you have time to make a patch for Wyze Cam v2? The latest firmware version is 4.9.8.1002.

@xraive, I took a look at iCamera for the v2 in this firmware version, but things are a bit different there (codebase looks similar but also a bit different) and it'll need more investigating.

**Comment by kohrar** [comment]
My PR should support the Wyze Cam v2 on the latest firmware as well. Please give any comments/feedback at #312 

I've also merged all my other PR fixes in my fork's master branch (https://github.com/kohrar/wz_mini_hacks) in case 0wish to try using the update script on my repo.

**Comment by xraive** [comment]
I don't know if this is the right place but now that I have my devices completely offline, how can I stop recording to MicroSD without the app? I am going to be recording over the network using motioneye or some other solution.

**Comment by jsolarin** [comment]
@xraive You can turn off recording in the /config/.user_config

recordType=3

https://github.com/gtxaspec/wz_mini_hacks/discussions/309

**Comment by xraive** [comment]
@jsolarin Thank you for this.  My apologies for not finding this.  I was searching to see if this was discussed in other issues, but I didn't see it.

**Comment by jsolarin** [comment]
@xraive I don't have any v2s to check with anymore but did you reboot the camera to reload the config?

If you have another v2 that you can put on the internet, you can watch that file and change the setting in the wyze app to see what it sets to.

**Comment by kohrar** [comment]
@xraive Can you please try setting it in `/configs/.parameters` instead and reboot to apply?

**Comment by xraive** [comment]
@kohrar looks like that did it.  Thanks @kohrar and @jsolarin.  The recording stopped after a reboot when I made the changes in the .parameters file.  So just to confirm the .parameters file is for v2 and the .user_config file is for v3?  Although the .user_config is also present in v2.

**Comment by kohrar** [comment]
@xraive Excellent!

That appears to be correct. The v2 firmware reads these settings from `.parameters`.

The v2's `.user_config` indeed exists and is written to by iCamera, but it appears to store the alarm/grid data (which is now stored in v3's `.grid_config` file).

**Comment by xraive** [comment]
@kohrar Would you consider patching WYZE Pan v1?

**Comment by kohrar** [comment]
> @kohrar Would you consider patching WYZE Cam Pan v1?

I was hoping the Pan v1 would be similar to the v2 firmware (version numbers are close, so I can just find the new addresses to patch), but the network handling code look a bit different. I don't have any of the pan cameras either to test with either, so I probably won't be able to make a patch for this anytime soon.

**Comment by xraive** [comment]
> > @kohrar Would you consider patching WYZE Cam Pan v1?
> 
> I was hoping the Pan v1 would be similar to the v2 firmware (version numbers are close, so I can just find the new addresses to patch), but the network handling code look a bit different. I don't have any of the pan cameras to test with either, so I probably won't be able to make a patch for this anytime soon.

No rush, if you need help with testing let me know.  I have a few of these devices.

**Comment by Bob-AOL** [comment]
Just finished this thread. Great work and amazing diligence among you guys. It's all magic to me.

Just curious if there may be any interest in reviewing #241 to possibly offer suggestions there for determining why only one of my three V2s accepted the new method of implementing wz_mini.  Our host has been pecking at it a little here and a little there, but I realize the priority level is pretty low in the bigger picture. 

I'm hardly a developer and only volunteered as a 'normal guy' guinea pig to test the new process. I was delighted that it went so smoothly with the first V2 and was excited about doing the other two, Unfortunately, we're still looking for some joy there.

TIA for any suggestions anyone may care to offer.

**Comment by rca** [comment]
One quick suggestion I can make is to format your SD Card and check, double check, and triple check that you are formatting it as MSDOS.  Any other filesystem type will not work.

**Comment by Bob-AOL** [comment]
A review of #241 includes all the details associated with the issue to date.

**Comment by rocket59** [comment]
I upgrade my V3 camera to 4.36.9.139 and set up the hacks successfully with ENABLE_SELFHOSTED_MODE changed to true in the conf file. However when I SSHd into the camera and ran ps it did not show the patch as active. The process shown is

(iCamera) /bin/sh /opt/wz_mini/usr/bin/iCamera 

(and the RTSP stream drops off after about a couple  of hours with no internet access).  Have I missed something?

**Comment by rocket59** [comment]
Well 10 hours on and it's still up. My thanks to @kohrar and @gtxaspec for the work that's gone into the hack and the self hosting mode.

I have 2 questions:
(1) I'm getting a lag of about 3 seconds (in VLC on the high res settings). Is there any way to reduce this a bit?
(2) @kohrar originally built the patch for the RTSP beta firmware. Out of curiosity I was going to try it with the beta firmware. Does that mean I would still install the mini hacks in order to get SSH access to the camera to execute the patch?

**Comment by Bob-AOL** [comment]
Am I to understand all of this as a LAN only solution with zero integrated Wyze app functionality, or am I overlooking something obvious that also allows it to be easily implemented in a WLAN environment? And if the latter is indeed applicable, what are the methods of choice for portable access to camera locations for live aiming, maintenance, remote viewing, etc?

**Comment by capncybo** [comment]
@Bob-AOL Not sure I follow where you are becoming confused but, with the default config inside wz_mini.conf...
ENABLE_SELFHOSTED_MODE="false"
-You will maintain WYZE app functionality & have Remote WAN type streaming via the WYZE app.
But there are some minor-major network issues (Depending on your network) infrastructure.

If you change... 
ENABLE_SELFHOSTED_MODE="false"
Then it becomes a local IP-Camera RTSP network solution, which severs the WAN -WYZE app connections
Hence you will be responsible for configuring your own Remote access from the Outside WAN.
In my case my ASUS Router supports OpenVPN.
But WireGuard May also  be an option for you.
Less Obvious to some may be that without the WYZE app you might have difficulties changing certain camera settings.
However if you also enable...
WEB_SERVER_ENABLED="true"
&
Visit the IP-Address of the camera you can possibly now change certain camera settings via the web-interface.
Note: this is a fairly recent addition.
+
You can also likely change certain camera parameters via ssh & using scripting...
However depending on your knowledge level... some of this may be too advanced for some.

**Comment by Bob-AOL** [comment]
Thanks for the response and clarification. I'm familiar with access to and modification of the .conf file by the various means available. I just wasn't clear about any side-effects regarding portable wireless app access to the device once the Wyze tethering had been severed -- e.g. when climbing a ladder for live positioning of a camera or just having handy app access for general viewing.

So the choices for, shall we say, "live viewing via portable app" are toggle the SELFHOSTED_MODE to obtain Wyze functionality for such purposes, or effectively at least achieve the same basic viewing ability via third party IP Cam RTSP streaming app such as VLC or TinyCam while in an active SELFHOSTED state -- taking the existing network environment into consideration, for such activity, of course.

Is that basically correct?

**Comment by jsolarin** [comment]
@Bob-AOL basically. The problem this issue solves is when you block the Wyze camera from the internet, the Wyze iCamera binary was restarting the network stack. This was interrupting the rtsp stream even on the Wyze rtsp firmware. So by patching the binary, it will no longer restart network and freeze the video stream. The side effect is losing the Wyze app management and cloud features but a small price to pay to be able to use the cameras offline. 

In the discussions we found out how to configure the camera directly so the Wyze app isn't needed. As well, the web interface has a camera stream you can view as well if needed but an rtsp application like vlc or tinycam is better.

**Comment by Vendo232** [comment]
@kohrar applied yout icamera patch, rebooted, upgrade, when running patch_icamera.sh it returns "patch already applied..." and yet pihole is still showing that the camera is pinging api.wyzecam.com.

any idea why is it? V3 update to the latest 139 FW

**Comment by Vendo232** [comment]
thank you @kohrar for the patch, all the api pings are gone

**Comment by sugatam** [comment]
Any way to get a patch for the iCamera binary on a Pan V2? Just supporting firmware 4.49.1.47, which is the last version that wz_mini_hacks supports, will be sufficient.
